exporter/kafkaexporter/kafka_exporter.go

```diff
@@ -29,10 +28,193 @@ import (
+	producer    sarama.SyncProducer
+	logger      *zap.Logger
+}
+
+func (k *SyncKafkaProducer) NewKafkaProducer(config Config, logger *zap.Logger) error {
+	conf, err := NewProducerByMessage(config, logger)
+	if err != nil {
+		return err
+	}
+	k.KafkaConfig = conf
+	k.producer, err = sarama.NewSyncProducer(conf.addressList, k.KafkaConfig.config)
+	if err != nil {
+		fmt.Println("sarama.NewSyncProducer err", err)
+		return err
+	}
+	return nil
+}
+
+func (k *SyncKafkaProducer) SendMessages(msg []*sarama.ProducerMessage) error {
+	err := k.producer.SendMessages(msg)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (k *SyncKafkaProducer) Close() error {
+	err := k.producer.Close()
+	if err != nil {
+		fmt.Println("sarama.NewSyncProducer close err")
+		return err
+	}
+	return nil
+}
+
+type AsyncKafkaProducer struct {
+	KafkaConfig *KafkaConfig
+	producer    sarama.AsyncProducer
+	logger      *zap.Logger
+}
+
+func (k *AsyncKafkaProducer) Close() error {
+	k.producer.AsyncClose()
+	return nil
+}
+
+func (k *AsyncKafkaProducer) NewKafkaProducer(config Config, logger *zap.Logger) error {
+	conf, err := NewProducerByMessage(config, logger)
+	if err != nil {
+		return err
+	}
+	k.KafkaConfig = conf
+	k.Run()
+	k.logger = logger
+	return nil
+}
+
+func NewProducerByMessage(config Config, logger *zap.Logger) (*KafkaConfig, error) {
+	c := sarama.NewConfig()
+	// These setting are required by the sarama implementation.
+	c.Producer.Return.Successes = true
+	c.Producer.Return.Errors = true
+	// Wait only the local commit to succeed before responding.
+	c.Producer.RequiredAcks = config.Producer.RequiredAcks
+	// Because sarama does not accept a Context for every message, set the Timeout here.
+	c.Producer.Timeout = config.Timeout
+	c.Metadata.Full = config.Metadata.Full
+	c.Metadata.Retry.Max = config.Metadata.Retry.Max
+	c.Metadata.Retry.Backoff = config.Metadata.Retry.Backoff
+	c.Producer.MaxMessageBytes = config.Producer.MaxMessageBytes
+	c.Producer.Flush.MaxMessages = config.Producer.FlushMaxMessages
+
+	if config.ProtocolVersion != "" {
+		version, err := sarama.ParseKafkaVersion(config.ProtocolVersion)
+		if err != nil {
+			return nil, err
+		}
+		c.Version = version
+	}
+
+	if err := ConfigureAuthentication(config.Authentication, c); err != nil {
+		return nil, err
+	}
+
+	compression, err := saramaProducerCompressionCodec(config.Producer.Compression)
+	if err != nil {
+		return nil, err
+	}
+	c.Producer.Compression = compression
+	if err := ConfigureAuthentication(config.Authentication, c); err != nil {
+		return nil, err
+	}
+	return &KafkaConfig{
+		addressList: config.Brokers,
+		topic:       config.Topic,
+		config:      c,
+		logger:      logger,
+	}, nil
+}
+
+func (k *AsyncKafkaProducer) SendMessages(value []*sarama.ProducerMessage) error {
+	if k.producer == nil {
+		k.Run()
+	}
+	k.producer.Input() <- value[0]
+	return nil
+
+}
+func (k *AsyncKafkaProducer) Run() {
+	if k == nil || k.KafkaConfig == nil {
+		return
+	}
```

Comment:
Testing if `k` is nil here feels like an anti pattern and could lead to issues trying to diagnose problems with the `AsyncKafkaProducer`. It should panic if these aren't correctly set in the worst case (in the best case, return an error on creation).